### DISCUSSION TITLE
WLAN.isconnected() is not really connected to wanted router when using static IP

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#9560] [Using MicroPython] [answered]
I know it was a silly mistake, but I have am use to setting up static IPs for my ESP32 devices.  So when I tried Mike Teachman's publishThingspeak.py I put some extra lines in to put my ThingSpeak device on a static IP.

Like this:
```py
# connect the device to the WiFi network
wifi = network.WLAN(network.STA_IF)
wifi.active(True)
utime.sleep(0.5)
wifi.ifconfig(('192.168.10.80', '255.255.255.0', '192.168.10.1', '8.8.8.8'))
wifi.connect(WIFI_SSID, WIFI_PASSWORD)

# wait until the device is connected to the WiFi network
MAX_ATTEMPTS = 20
attempt_count = 0
while not wifi.isconnected() and attempt_count < MAX_ATTEMPTS:
    attempt_count += 1
    utime.sleep(1)

if attempt_count == MAX_ATTEMPTS:
    print('could not connect to the WiFi network')
    sys.exit()

print ('connected to WiFi')
print (' network config:', wifi.ifconfig())
```
When I changed to a different router on a different network isconnected() said is was connected but
apparently to 192.168.10.80 not the actual router I was trying to connect to.

Reading the docs:

> WLAN.isconnected()
    In case of STA mode, returns True if connected to a WiFi access point and has a valid IP address.

It might be difficult for isconnected() to recognise the user mistake.  Or is there a better way to setup static IPs?

**Comment by puppet13th** [comment, marked as answer]
`wifi.ifconfig(('192.168.10.80', '255.255.255.0', '192.168.10.1', '8.8.8.8'))`
192.168.10.80 -> client ip address(esp32)
255.255.255.0 -> netmask
192.168.10.1 -> gateway(should be router)
8.8.8.8 -> dns

for client static ip address you have 2 options:
1. set static address from client side(like what you are doing now)
2. set dhcp static address from router side(add client mac address(esp32) to router dhcp static lease)

for option 2,you should not use this line:
`wifi.ifconfig(('192.168.10.80', '255.255.255.0', '192.168.10.1', '8.8.8.8'))`

**Reply by davefes** [reply]
I am happy using option 1.

My main point is that I don't think that the documentation statement is true, because it looks like I managed the confuse the ESP32 WiFi setup by my mistake.

**Reply by puppet13th** [reply]
```
WLAN.isconnected()
In case of STA mode, returns True if connected to a WiFi access point and has a valid IP address.
```
so there is 2 things:
1. connected to access point
2. valid ip address

in your setup:
1. connected to access point -> True
2. valid ip address -> True (static from ifconfig setting)
so calling WLAN.isconnected() should return True

in other condition when router did not assign ip address to client and client does not have static ip address setting:
1. connected to access point -> True
2. valid ip address -> False
in this condition calling WLAN.isconnected() should return False

valid ip address = static ip address(WLAN.ifconfig) or got dhcp ip address from router.
example code:
```
import network
from binascii import hexlify
wlan = network.WLAN(network.STA_IF); wlan.active(True)
wlan.connect('test','password')
print(wlan.isconnected())
print(wlan.ifconfig())
print(hexlify(wlan.config('mac')))
```
output:
```
>>> %Run -c $EDITOR_CONTENT
False
('0.0.0.0', '0.0.0.0', '0.0.0.0', '0.0.0.0')
b'6055f9717114'
```
![image](https://user-images.githubusercontent.com/48564964/194749359-9374009d-2d8e-4b5d-ba19-0236c059ce41.png)
esp32 connected to access point but did not have static ip address or got ip address from router.

**Reply by davefes** [reply]
puppet13,
I understand now what you are saying that the documentation states.  Both conditions were true but because of my error I ended up with a non-functional wifi connection.

Thank you for taking the time to explain.
